src/Jackalope/Version/VersionHandler.php

```diff
@@ -0,0 +1,66 @@
+<?php
+
+namespace Jackalope\Version;
+
+use Jackalope\Node;
+use Jackalope\Transport\TransportInterface;
+use PHPCR\UnsupportedRepositoryOperationException;
+
+class VersionHandler
+{
+    const MIX_VERSIONABLE = 'mix:versionable';
+    const MIX_SIMPLE_VERSIONABLE = 'mix:simpleVersionable';
+
+    /**
+     * @var TransportInterface
+     */
+    private $transport;
+
+    public function __construct(TransportInterface $transport)
+    {
+        $this->transport = $transport;
+    }
+
+    public static function addVersionProperties(Node $node, TransportInterface $transport)
+    {
+        // TODO solve without reflection
+        $reflection = new \ReflectionMethod(get_class($transport), 'storeNode');
+        $reflection->setAccessible(true);
+
+        // TODO move this to the version handler
+        $session = $node->getSession();
+        $rootNode = $session->getRootNode();
+
+        $node->setProperty('jcr:isCheckedOut', true);
+
+        if (!$rootNode->hasNode('jcr:system/jcr:versionStorage')) {
+            if (!$rootNode->hasNode('jcr:system')) {
+                $rootNode->addNode('jcr:system');
+            }
+            $rootNode->addNode('jcr:system/jcr:versionStorage');
+        }
+
+        $versionStorageNode = $rootNode->getNode('jcr:system/jcr:versionStorage');
+        if (!$versionStorageNode->hasNode($node->getIdentifier())) {
+            $versionStorageNode->addNode($node->getIdentifier(), 'nt:versionHistory');
+        }
+        $versionHistory = $versionStorageNode->getNode($node->getIdentifier());
+        $versionHistory->setProperty('jcr:versionableUuid', $node->getIdentifier());
+
+        // TODO Set jcr:copiedFrom if needed
+
+        $versionHistory->addNode('jcr:versionLabels', 'nt:versionLabels');
+        $rootVersion = $versionHistory->addNode('jcr:rootVersion', 'nt:version');
+        $reflection->invoke($transport, $rootVersion->getPath(), $rootVersion->getProperties());
```

Comment:
Ok, so keep the `VersionHandler`, but inject the `ObjectManager` to it? I tried to do that in the first place, but I didn't find a way to this easily, ergo without breaking the current architecture. The only option I've seen so far is to add something like the following after line https://github.com/jackalope/jackalope/blob/be3d34bd63cdd9200d9ec469dee0bc0cf671ccda/src/Jackalope/ObjectManager.php#L164

```
if ($this->transport implements MyVersioningInterface){
    $this->transport->setVersionHandler(new VersionHandler($this));
}
```

Of course I would have to think about a better name for this interface :smile: